modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpCache.java

```diff
@@ -55,6 +56,22 @@ AbstractResponse get(InetAddress ip, String databasePath) {
         return cache.get(cacheKey);
     }
 
+    public int purgeCacheEntriesForDatabase(Path databaseFile) {
+        String databasePath = databaseFile.toString();
+        int counter = 0;
+        for (CacheKey key : cache.keys()) {
+            if (key.databasePath.equals(databasePath)) {
```

Comment:
Is this synchronized in any way? Is it possible that we will add new key during invalidation of old ones?

Reply:
Good catch. This does need some synchronization, otherwise there may be a chance that keys are added for a database to be deleted.

Reply:
I added reference counting to `DatabaseReaderLazyLoader`, so that keys for the old databases are only removed immediately after the update if there are no usages from geoip processor instances or when the last geoip processor has completed doing geo enrichment.

This also addresses another issue; when a geoip processor gets a `DatabaseReaderLazyLoader` instance and then a database update occurs for the same database then the maxmind db reader (which is wrapped by `DatabaseReaderLazyLoader`) may fail with an 'com.maxmind.db.ClosedDatabaseException: The MaxMind DB has been closed.' error. The reference counting avoids closing the reader if it is being used.